Rh button group

HadassahYelenik opened the pull request:
## What I did

1. Created a new rh-button-group web component.

2. Added styling (rh-button-group.css) to handle button layout and spacing.

3. Built a demo page (demo/index.html) to showcase component usage.

4. Implemented TypeScript functionality (rh-button-group.ts) for component behavior.

## Testing Instructions

1. Open elements/rh-button-group/demo/index.html in a browser.

2. Verify that the buttons display correctly in a group layout.

3. Click the buttons to ensure any interactions work as expected.

4. Optional: Integrate the component into a test page in the repo to see it in context.

## Notes to Reviewers
1. This is a new component, so no breaking changes to existing functionality.

2. Styles and behavior are confined to rh-button-group.

3. Feedback welcome on styling, accessibility, and any API improvements.

bennypowers commented:
@adamjohnson please give this a ux review

@coreyvickery you mentioned in internal chat some design guidelines, if it's not too much trouble, please add (publically accessible) links to those design files here

coreyvickery commented:
I just grabbed [spacing guidance](https://www.patternfly.org/components/button/design-guidelines#spacing) from PatternFly 6.

And [here](https://www.patternfly.org/tokens/all-patternfly-tokens) are their spacer token values.

HadassahYelenik commented:
relates to issue #2558

 ## What I did
> 1. Created a new rh-button-group web component.
> 2. Added styling (rh-button-group.css) to handle button layout and spacing.
> 3. Built a demo page (demo/index.html) to showcase component usage.
> 4. Implemented TypeScript functionality (rh-button-group.ts) for component behavior.
> 
> ## Testing Instructions
> 1. Open elements/rh-button-group/demo/index.html in a browser.
> 2. Verify that the buttons display correctly in a group layout.
> 3. Click the buttons to ensure any interactions work as expected.
> 4. Optional: Integrate the component into a test page in the repo to see it in context.
> 
> ## Notes to Reviewers
> 1. This is a new component, so no breaking changes to existing functionality.
> 2. Styles and behavior are confined to rh-button-group.
> 3. Feedback welcome on styling, accessibility, and any API improvements.

coreyvickery commented:
@bennypowers Should any docs updates be handled in this issue?

bennypowers commented on elements/rh-button-group/rh-button-group.ts:
```diff
@@ -0,0 +1,111 @@
+  private _buttons!: (HTMLButtonElement | RhButton)[];
+
+  override firstUpdated() {
+    if (!this._buttons?.length) {
```
this will likely fail in SSR, since we don't have access to children in the SSR nodejs context

bennypowers commented on elements/rh-button-group/rh-button-group.ts:
```diff
@@ -0,0 +1,111 @@
+  /**
+   * Setup roving tabindex for toolbar
+   */
+  #setupToolbar() {
```
use RovingTabindexController for this

bennypowers commented on elements/rh-button/rh-button.ts:
```diff
@@ -13,6 +13,7 @@ import { InternalsController } from '@patternfly/pfe-core/controllers/internals-
 
 import styles from './rh-button.css';
 
+
```
revert whitespace changes

bennypowers commented:
@coreyvickery consider these examples (final attributes may change):

```html
<rh-button-group>
  <rh-button>Shutdown</rh-button>
  <rh-button>Restart</rh-button>
  <rh-button>Logout</rh-button>
</rh-button-group>


<rh-button-group variant="radio">
  <rh-button>Dark</rh-button>
  <rh-button>Light</rh-button>
  <rh-button>System</rh-button>
</rh-button-group>


<rh-button-group>
  <rh-menu-dropdown>...</rh-menu-dropdown>
  <rh-button>Manage Account</rh-button>
</rh-button-group>
```

button group is a container for buttons

coreyvickery commented:
Got it. Docs are in progress! 🙂

HadassahYelenik commented:
@coreyvickery thanks for letting me know, I’m looking into the issue now and will update the branch once it’s fixed.

HadassahYelenik commented:
> There are many unrelated changes here that must be reverted. perhaps the result of a bad merge?
@bennypowers you were right, there was a bad merge that brought in many unrelated changes. I've cleaned up the branch history by creating a new one from main and cherry-picking only the relevant commits for this feature. The PR should now be clean and only contain the intended changes. Please let me know if there is anything else.


> @HadassahYelenik Some changes requested:
> 
> * [ ]  Space between buttons in a row is `--rh-space-lg` (16px)
> * [ ]  Space between buttons in a column is `--rh-space-md` (8px)
> * [ ]  For more information, please read through [the docs](https://www.figma.com/proto/kRy2Ojl2esFL58SwZxN1Ge/Spacing-updates?page-id=60%3A12030&node-id=60-12031&viewport=114%2C138%2C1&t=pDruD2aP3bFEoRqg-1&scaling=scale-down&content-scaling=fixed)
@coreyvickery I've updated the component's styling to meet the spacing requirements.


> > @adamjohnson Docs are [here](https://www.figma.com/proto/kRy2Ojl2esFL58SwZxN1Ge/Spacing-updates?page-id=60%3A12030&node-id=60-12031&viewport=114%2C138%2C1&t=pDruD2aP3bFEoRqg-1&scaling=scale-down&content-scaling=fixed) if you'll be working on them.
> 
> @HadassahYelenik should be able to handle coding the docs up. Let us know if you have any questions. Reference other elements/files ala elements/rh-_/docs/_ to see the coding patterns we use.
@adamjohnson @coreyvickery I need permission to veiw [this](https://www.figma.com/proto/kRy2Ojl2esFL58SwZxN1Ge/Spacing-updates?page-id=60%3A12030&node-id=60-12031&viewport=114%2C138%2C1&t=pDruD2aP3bFEoRqg-1&scaling=scale-down&content-scaling=fixed)

HadassahYelenik commented:
> @HadassahYelenik Docs attached.
> 
> [Button group docs.fig.zip](https://github.com/user-attachments/files/22604419/Button.group.docs.fig.zip)

Thank you @coreyvickery.

bennypowers commented on elements/rh-button-group/rh-button-group.css:
```diff
@@ -3,8 +3,5 @@
-    --rh-space-md,
-    16px
-  );
+  margin-bottom: var(--rh-space-md, 8px);
```
```suggestion
  margin-block-end: var(--rh-space-md, 8px);
```

https://developer.mozilla.org/en-US/docs/Glossary/Logical_properties

bennypowers commented on elements/rh-button-group/test/rh-button-group.spec.ts:
```diff
@@ -0,0 +1,75 @@
+      await element.updateComplete;
+
+      const buttons = element.querySelectorAll('rh-button');
+      expect(buttons).to.have.length(2);
```
we can remove this test, it doesn't make any assertions about our element's public apis, state, or observable side effects

HadassahYelenik commented on elements/rh-button-group/rh-button-group.ts:
```diff
@@ -0,0 +1,72 @@
+  @queryAssignedElements({ flatten: true })
+  private _buttons!: (HTMLButtonElement | RhButton)[];
+
+  private _rovingController?: RovingTabindexController<HTMLButtonElement | RhButton>;
```
Great suggestion — I’ve updated it so the controller is always created and just noops when role !== 'toolbar'.
This feels cleaner than the conditional creation and aligns with how rh-tabs and rh-menu handle their controllers. Thanks for the guidance!

coreyvickery commented:
@HadassahYelenik Below is the image for the `All elements` page. File path would be: `/elements/button-group/screenshot.png`

The body text is:

> Organize multiple related buttons into a single collection

<img width="293" height="34" alt="screenshot" src="https://github.com/user-attachments/assets/ea9088d3-1f2f-4a7f-b139-8f7245b4483f" />

cc @adamjohnson If you need to do this instead.

adamjohnson commented on elements/rh-button-group/rh-button-group.ts:
```diff
@@ -0,0 +1,70 @@
+/**
+ * A button group organizes multiple related buttons into a single collection,
+ * providing visual grouping and coordinated keyboard navigation behavior.
+ *
```
```suggestion
 *
 * @summary Organize multiple related buttons into a single collection
 * 
```

@HadassahYelenik Regarding [this question](https://github.com/RedHat-UX/red-hat-design-system/pull/2605#issuecomment-3435732740), this is how you add a short description for an element on the "All Elements" page.

adamjohnson replied:
Here's an example with [rh-disclosure](https://github.com/RedHat-UX/red-hat-design-system/blob/main/elements/rh-disclosure/rh-disclosure.ts#L26) too for reference. Most/all of our elements should have a `@summary`.

HadassahYelenik replied:
Thank you @adamjohnson for clarifying how this works and for your guidance!

bennypowers replied:
documentation for all of the fields in custom elements manifest is available in the [schema](https://github.com/webcomponents/custom-elements-manifest/blob/main/schema.d.ts)

see [cem documentation](https://bennypowers.dev/cem/docs/commands/generate/) for notes on how to add docs inline

adamjohnson commented on elements/rh-button-group/docs/10-style.md:
Looks like every image in this file references a `.png`. We generally export these assets as SVG images (eg: `example.svg`) and link them ala `src="../button-group-anatomy.svg"`. 

Exporting them as SVG files ensures they are crisp, vector images no matter what screen density (eg "Retina") they are viewed with.

Kindly update the images and their references in this file and the other `docs/*.md` files to `*.svg`.

adamjohnson commented on elements/rh-button-group/docs/20-guidelines.md:
As above, update these images from `*.png` to be SVG files and then update the corresponding `src` attributes.

adamjohnson commented on elements/rh-button-group/docs/00-overview.md:
```diff
@@ -0,0 +1,10 @@
+
+<div id="overview-image-description" class="visually-hidden">
+  Image of a button group element
+</div>
```
```suggestion
```

This can be removed. It's an old artifact from before when we didn't have knobs (aka the interactive demo on the Overview page).

adamjohnson commented on elements/rh-button-group/docs/10-style.md:
```diff
@@ -0,0 +1,68 @@
+When button groups are vertical, there is less padding.
+
+<uxdot-example color-palette="lightest" width-adjustment="680px">
+  <img alt="Diagram showing spacing for vertical and horizontal button groups." 
```
Rewrite this alt text to specifically say what the image communicates. Call out which buttons, links and orientations have which padding (eg 8px, 16px, 32px, or 48px). Concurrently, try to keep it as concise as possible.

adamjohnson commented on elements/rh-button-group/demo/index.html:
```diff
@@ -0,0 +1,17 @@
+<rh-button-group role="group">
```
Instead of adding `role="group"` (and with regard to the considerations in the discussion a few lines below this one), could we handle this with the InternalsController? We generally use the InternalsController to define semantics like this. eg [rh-navigation-link](https://github.com/RedHat-UX/red-hat-design-system/blob/main/elements/rh-navigation-link/rh-navigation-link.ts#L30). 

If we need to change the role, [rh-icon has an example of that](https://github.com/RedHat-UX/red-hat-design-system/blob/main/elements/rh-icon/rh-icon.ts#L161-L164). That said, I don't know if that will be necessary if we decide to make `<rh-button-group-toolbar>` (or whatever it would be named).

adamjohnson commented on elements/rh-button-group/README.md:
```diff
@@ -0,0 +1,13 @@
+  <button type="button">Save</button>
+  <button type="button">Cancel</button>
+  <button type="button">Delete</button>
+</rh-button-group>
```
This example will likely change as we make API changes as discussed below.

adamjohnson commented on .changeset/rh-button-group-component.md:
```diff
@@ -0,0 +1,15 @@
+  <rh-button>Save</rh-button>
+  <rh-button>Cancel</rh-button>
+  <rh-button>Delete</rh-button>
+</rh-button-group>
```
This example will likely change as we make API changes as discussed below.

adamjohnson commented on elements/rh-button-group/docs/10-style.md:
```diff
@@ -0,0 +1,68 @@
+</figure>
+
+### Orientation
+
```
Also, if the docs are calling out "vertical orientations", that likely means we'll have to properly handle `aria-orientation` and include demos for a vertical orientation too. (See discussion above).

adamjohnson reviewed (changes requested):
Some things to think about:

1. Possible API changes around `role="group"` and `role="toolbar"`
2. Updates to docs imagery and content
3. Possible additon of an orientation demo
4. Possible changes to tests.

Also, instead of having the base branch based on `main` (or `RedHat-UX:main`), it should be based off of `staging/eevee`.

adamjohnson commented on elements/rh-button-group/demo/horizontal.html:
This file is pretty similar to the `index.html` demo. I would suggest removing the first button group (save, cancel, more) and keeping the second button group. Then, rename this file to `icons.html`.

adamjohnson commented on elements/rh-button-group/docs/button-group-anatomy.svg:
@coreyvickery I see the "Color Scheme" picker in the images and elsewhere in the docs. In this case, that element uses [radio inputs](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/input/radio) and using a button group would not be semantically appropriate / inaccessible because buttons don't have a checked/selected state. 

My recomendation would be to remove all color scheme pickers from these docs + images.

HadassahYelenik replied:
@coreyvickery When you have a chance, could you please send the updated docs as a zip file?

adamjohnson commented on elements/rh-button-group/rh-button-group.ts:
```diff
@@ -0,0 +1,67 @@
+  #internals = InternalsController.of(this, { role: 'group' });
+
+  /** Sets the orientation of the button group */
+  @property({ reflect: true }) orientation: 'horizontal' | 'vertical' = 'horizontal';
```
`aria-orientation` does not apply to elements with `role="group"`. It _does_ apply to elements with `role="toolbar"`. [W3C docs](https://w3c.github.io/aria/#group).

So, should we decide to support `role="toolbar"` (or a wrapper custom element for rh-button-group), this property would go on that custom element then apply the actual `aria-orientation` property in that element's template/HTML. 

If we are not going to support the toolbar role, we can keep this property and note in the docs on line 31 something like "Visually sets the orientation..."

bennypowers replied:
From docs. Toolbar is:

> A collection of commonly used function buttons or controls represented in compact visual form.
> 
> The toolbar is often a subset of functions found in a menubar, designed to reduce user effort in using these functions. Authors MUST supply a label on each toolbar when the application contains more than one toolbar.
> 
> Authors MAY manage focus of descendants for all instances of this role, as described in Managing Focus.
> 
> Elements with the role toolbar have an implicit aria-orientation value of horizontal.

Group is:
 
> A set of user interface objects that is not intended to be included in a page summary or table of contents by assistive technologies.
> 
> Contrast with region, which is a grouping of user interface objects that will be included in a page summary or table of contents.
> 
> Authors SHOULD use a group to form a logical collection of items in a widget, such as children in a tree widget forming a collection of siblings in a hierarchy. However, when a group is used in the context of a listbox, for example, authors MUST limit its children to option elements. Therefore, proper handling of group by authors and assistive technologies is determined by the context in which it is provided.
> 
> Authors MAY nest group elements. If a section is significant enough to warrant inclusion in the web page's table of contents, the author SHOULD assign it a role of region or a standard landmark role.

While group seems to me applicable, seems like button group is more like a toolbar

@hellogreg wdyt?

HadassahYelenik replied:
@BennyPowers @adamjohnson @hellogreg
If we decide to create a separate `<rh-button-group-toolbar>` wrapper, as Adam suggested [here](https://github.com/RedHat-UX/red-hat-design-system/pull/2605#discussion_r2492032751), I want to make sure `<rh-button-group>` works well on its own as a single group. Is there anything I should adjust now to make it ready for that wrapper?

HadassahYelenik replied:
Thanks for clarifying! I understand that aria-orientation only applies to elements with `role="toolbar"`, so that will go on `<rh-button-group-toolbar>`. For `<rh-button-group>` itself, should I remove the current visual-only orientation property and leave it for the toolbar wrapper, or keep it for single-group layout as it is now?

coreyvickery commented:
> Hi @coreyvickery, just wanted to check in on the `<rh-button-group>` docs. Whenever they’re ready, I can implement the updates—thanks so much!

@HadassahYelenik Attached.

[Button group docs 11-21.fig.zip](https://github.com/user-attachments/files/23682518/Button.group.docs.11-21.fig.zip)

adamjohnson commented on elements/rh-button-group/demo/icons.html:
```diff
@@ -0,0 +1,11 @@
+  <rh-button variant="secondary" icon="copy" label="Copy"></rh-button>
+  <rh-button variant="secondary" icon="undo" label="Undo"></rh-button>
+  <rh-button variant="secondary" icon="trash" label="Delete"></rh-button>
+</rh-button-group>
```
```suggestion
<rh-button-group>
  <rh-button variant="secondary" icon="edit" accessible-label="Edit"></rh-button>
  <rh-button variant="secondary" icon="copy" accessible-label="Copy"></rh-button>
  <rh-button variant="secondary" icon="undo" accessible-label="Undo"></rh-button>
  <rh-button variant="secondary" icon="trash" accessible-label="Delete"></rh-button>
</rh-button-group>
```

We've updated this API recently and are going to deprecate `label` when `staging/eevee` releases. See #2762 for more info.

adamjohnson reviewed (approved):
Layouts Group Toggles Magnificently (LGTM!)

> Are there any remaining updates needed for the docs?

We can handle those in a separate docs PR. I've [created an issue](https://github.com/RedHat-UX/red-hat-design-system/issues/2769) to track that. Great work here!

Go ahead and click "Squash and merge" below to merge this PR. 👏

bennypowers reviewed (changes requested):
a few nits, otherwise this is just about ready to go

bennypowers reviewed (approved):
Linearly Grouped Toggle Mechanisms
